File: mixin-base/src/main/java/net/yakclient/mixin/base/internal/methodadapter/tree/TreeBEPatternMatcher.java
package net.yakclient.mixin.base.internal.methodadapter.tree;

import net.yakclient.mixin.base.internal.bytecode.ByteCodeUtils;
import net.yakclient.mixin.base.internal.instruction.Instruction;
import org.objectweb.asm.tree.AbstractInsnNode;
import org.objectweb.asm.tree.InsnList;

/**
 * Defines the pattern for inserting mixins before the last return of
 * the specified instructions.
 * <p>
 * Type: Before end
 *
 * @author Durgan McBroom
 */
public class TreeBEPatternMatcher extends TreeMixinPatternMatcher {
    public TreeBEPatternMatcher(Instruction instructions) {
        super(instructions);
    }

    @Override
    public void transform(InsnList insn) {
        AbstractInsnNode lastReturn = null;
        for (AbstractInsnNode node : insn) {
            if (ByteCodeUtils.isReturn(node.getOpcode())) lastReturn = node;
        }
        this.insertInsn(lastReturn, insn);
    }
}

File: base/src/main/kotlin/net/yakclient/mixins/base/internal/methodadapter/tree/TreeBEPatternMatcher.kt
package net.yakclient.mixins.base.internal.methodadapter.tree

import net.yakclient.mixins.base.internal.bytecode.ByteCodeUtils.isReturn
import net.yakclient.mixins.base.internal.instruction.Instruction
import org.objectweb.asm.tree.AbstractInsnNode
import org.objectweb.asm.tree.InsnList

/**
 * Defines the pattern for inserting mixins before the last return of
 * the specified instructions.
 *
 *
 * Type: Before end
 *
 * @author Durgan McBroom
 */
class TreeBEPatternMatcher(
    instructions: Instruction
) : TreeMixinPatternMatcher(instructions) {

    override fun transform(insn: InsnList) {
        var lastReturn: AbstractInsnNode? = null
        for (node in insn) {
            if (isReturn(node.opcode)) lastReturn = node
        }
        this.insertInsn(lastReturn, insn)
    }
}